### ISSUE TITLE
is_namespaced is not always discovered correctly

### ISSUE THREAD

**Issue by amadev** [nolar/kopf#613]
## Long story short

In https://github.com/nolar/kopf/blob/214310c8f678fad5e267aacfbbc8acdadb557d9c/kopf/clients/discovery.py#L17
context._discovered_resources may not contain all necessary info. For example, if Resource(group='', version='v1', plural='secrets')
is discovered first, it's filled with 'v1': {}. Thus, node resource discovered wrongly after that.

## Description

[2020-12-14 17:26:46,627] kopf.reactor.running [ERROR   ] Root task 'watcher of nodes' is failed: 404, message='Not Found', url=URL('https://172.16.244.44:6443/api/v1/namespaces/openstack/nodes')
Traceback (most recent call last):
  File "/home/amadev/m/openstack-controller/.tox/dev/lib/python3.7/site-packages/kopf/reactor/running.py", line 438, in _root_task_checker
    await coro
  File "/home/amadev/m/openstack-controller/.tox/dev/lib/python3.7/site-packages/kopf/reactor/queueing.py", line 128, in watcher
    async for raw_event in stream:
  File "/home/amadev/m/openstack-controller/.tox/dev/lib/python3.7/site-packages/kopf/clients/watching.py", line 71, in infinite_watch
    async for raw_event in stream:
  File "/home/amadev/m/openstack-controller/.tox/dev/lib/python3.7/site-packages/kopf/clients/watching.py", line 110, in streaming_watch
    async for raw_event in stream:
  File "/home/amadev/m/openstack-controller/.tox/dev/lib/python3.7/site-packages/kopf/clients/watching.py", line 128, in continuous_watch
    items, resource_version = await fetching.list_objs_rv(resource=resource, namespace=namespace)
  File "/home/amadev/m/openstack-controller/.tox/dev/lib/python3.7/site-packages/kopf/clients/auth.py", line 46, in wrapper
    return await fn(*args, **kwargs, context=context)
  File "/home/amadev/m/openstack-controller/.tox/dev/lib/python3.7/site-packages/kopf/clients/fetching.py", line 99, in list_objs_rv
    response.raise_for_status()
  File "/home/amadev/m/openstack-controller/.tox/dev/lib/python3.7/site-packages/aiohttp/client_reqrep.py", line 946, in raise_for_status
    headers=self.headers)
aiohttp.client_exceptions.ClientResponseError: 404, message='Not Found', url=URL('https://172.16.244.44:6443/api/v1/namespaces/openstack/nodes')

## Environment

kopf version: master
(dev) ~/m/openstack-controller$ kubectl version
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.2", GitCommit:"52c56ce7a8272c798dbc29846288d7cd9fbae032", GitTreeState:"clean", BuildDate:"2020-04-16T11:56:40Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18+", GitVersion:"v1.18.8-mirantis-1", GitCommit:"43e7409273737fcaa74b2c010bdb5ddc24c1701c", GitTreeState:"clean", BuildDate:"2020-08-19T03:01:58Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
(dev) ~/m/openstack-controller$ python --version
Python 3.7.7
(dev) ~/m/openstack-controller$ uname -a
Linux flashman 5.3.0-51-generic #44~18.04.2-Ubuntu SMP Thu Apr 23 14:27:18 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

**Comment by nolar** [comment]
@amadev Indeed. Thanks for the report.

Can you please send a PR with that fix in the mentioned commit?  You can skip writing the unit-tests for the fix — see the notes below.

And a little request: when you will be sending the PR, can you please base it on the branch `release/0.28`, not on `master`? It will help me to release a hotfix (0.28.3) without back-porting the code manually (and thus rewriting your commits).

---

**For the record and for history:** 

To reproduce:

```python
import kopf

@kopf.on.event('', 'v1', 'secrets')
@kopf.on.event('', 'v1', 'nodes')
def fn(**kwargs):
    pass
```

I've put this print before the if condition:

```python
    k = {k: list(v) for k, v in context._discovered_resources.items()}
    print(f"{resource.name} ??? {context._discovered_resources.keys()} ??? {k}")
```

And it shows this:

```
nodes ??? dict_keys([]) ??? {}
secrets ??? dict_keys(['v1']) ??? {'v1': []}
nodes ??? dict_keys(['v1']) ??? {'v1': ['bindings', 'componentstatuses', 'configmaps', 'endpoints', 'events', 'limitranges', 'namespaces', 'namespaces/finalize', 'namespaces/status', 'nodes', 'nodes/proxy', 'nodes/status', 'persistentvolumeclaims', 'persistentvolumeclaims/status', 'persistentvolumes', 'persistentvolumes/status', 'pods', 'pods/attach', 'pods/binding', 'pods/eviction', 'pods/exec', 'pods/log', 'pods/portforward', 'pods/proxy', 'pods/status', 'podtemplates', 'replicationcontrollers', 'replicationcontrollers/scale', 'replicationcontrollers/status', 'resourcequotas', 'resourcequotas/status', 'secrets', 'serviceaccounts', 'serviceaccounts/token', 'services', 'services/proxy', 'services/status']}
secrets ??? dict_keys(['v1']) ??? {'v1': ['bindings', 'componentstatuses', 'configmaps', 'endpoints', 'events', 'limitranges', 'namespaces', 'namespaces/finalize', 'namespaces/status', 'nodes', 'nodes/proxy', 'nodes/status', 'persistentvolumeclaims', 'persistentvolumeclaims/status', 'persistentvolumes', 'persistentvolumes/status', 'pods', 'pods/attach', 'pods/binding', 'pods/eviction', 'pods/exec', 'pods/log', 'pods/portforward', 'pods/proxy', 'pods/status', 'podtemplates', 'replicationcontrollers', 'replicationcontrollers/scale', 'replicationcontrollers/status', 'resourcequotas', 'resourcequotas/status', 'secrets', 'serviceaccounts', 'serviceaccounts/token', 'services', 'services/proxy', 'services/status']}
```

The 2nd line is the problem: it should not be an empty list and should be either populated already, or "v1" should be absent  (both cases are fine) — the same as it already happens a moment later in the 3rd line.

Looks like an async synchronisation issue. The if-lock-if (or "fast-if→lock→sure-if") pattern is used in order to make a quick check without heavy locks: it goes for a lock only if we "probably" are going to modify the structure; and under the lock, it double-checks again to change from "probably" to "certainly", and to make sure that is on us, not on other tasks.

However, the condition of the "fast-if" is wrong here. It checks for the top-level API group/version existence only. When one of the tasks populates the cache, it first creates an empty key-value pair, and then makes a slow API request, and then updates it with more resources. If that would be one atomic or almost-atomic operation (without the slow network activity in between), all would be fine or almost-fine. But it is not: between these two cache modifications, some other tasks notice that the `"v1"` key is present, assume (wrongly) that the whole structure is present, and skip the discovery completely (even the locking part).

The solution suggested in https://github.com/amadev/kopf/commit/801ecbd2f20c5346e66dd9e34e179a9ad3240c38 seems sufficient (but no PR?).

**Worth noting** that any fix, so as the whole concept of discovery as it is now in 0.28.x, will be gone after #600 and in the release 0.29. There, a completely different way of discovering the resources and their properties is implemented.

**Comment by nolar** [comment]
PS: I forgot to mention — please base the fix on `release/0.28` instead of `master`, **AND** also send the PR to that branch, not  to `master`.

**Comment by amadev** [comment]
@nolar Thanks for the detailed answer.

**Comment by nolar** [comment]
Released as `kopf==0.28.3` ([change notes](https://github.com/nolar/kopf/releases/tag/0.28.3)).

I presume it is fixed. If not — please reopen.

@amadev And thanks for the contribution!
